File: scripts/okf-prompt-lookup.py
#!/usr/bin/env python3
"""okf-prompt-lookup - score knowledge/index.md lines against a free-form prompt.

Standalone. No Headroom dependency. Stdlib only. Sibling to okf-index-lookup.py
(which does single-term substring); this one does prompt-tokenised overlap.

Usage:
    python scripts/okf-prompt-lookup.py "rotate a leaked GitHub token"
    python scripts/okf-prompt-lookup.py --limit 5 --min-score 2 "..."
    echo '{"prompt":"..."}' | python scripts/okf-prompt-lookup.py

Reads prompt from argv or JSON-stdin (e.g. CC UserPromptSubmit hook). Returns
top-N concept-file paths from `knowledge/index.md` whose lines best match the
prompt's content tokens.

Configuration (precedence, lowest to highest):
  1. Built-in defaults (limit=3, min_score=1, etc.)
  2. scripts/okf-lookup.config.json — every knob documented inline
  3. OKF_LOOKUP_<KEY> env vars — e.g. OKF_LOOKUP_LIMIT=5
  4. CLI flags — --limit, --min-score, --format, --root, --no-glossary, etc.

Behaviour:
- Tokenise prompt: lowercase, strip punctuation, drop stopwords + short tokens.
- For each line in `index.md`, score = sum of distinct prompt tokens present.
- Optional +1 boost for lines linking to a runbook (--boost-runbooks).
- Extract every `[label](./path.md)` link target on qualifying lines.
- Rank link targets by (score desc, first-seen order). Return top-N.
- Exit 1 if zero hits (hook caller decides whether to suppress).

Wire into agent hooks via JSON-stdin (CC) or argv (other agents). ~30ms on
1300-line index. Deterministic. No network, no embeddings, no LLM.
"""
from __future__ import annotations

import argparse
import json
import os
import re
import sys
from pathlib import Path

LINK_RE = re.compile(r"\[[^\]]+\]\(([^)]+)\)")
TOKEN_RE = re.compile(r"[a-z0-9][a-z0-9\-_]+")

# English fillers + agent-prompt noise. Removed from token set before scoring.
STOPWORDS = frozenset("""
the a an and or but if then else when while for from to of in on at by with
about into over under across this that these those it its is are was were be
been being have has had do does did will would should could may might must
can shall i me my we us our you your they them their he she his her
please thanks thank yes no ok okay sure can pls plz just want need help
how what why where which who whose make let lets set get got make have take
""".split())

# Oriz-specific stopwords. Suppressed via config.domain_stopwords_enabled=false
# when the script is forked for a non-oriz repo.
DOMAIN_STOPWORDS = frozenset("""
oriz knowledge agent agents rule rules file files repo repos
""".split())

DEFAULT_CONFIG = {
    "limit": 3,
    "min_token_length": 3,
    "min_score": 1,
    "max_line_preview": 120,
    "index_path": "knowledge/index.md",
    "format": "reminder",
    "extra_stopwords": [],
    "domain_stopwords_enabled": True,
    "include_glossary": True,
    "boost_runbooks": False,
}


def load_config(script_dir: Path) -> dict:
    """Load config from JSON file next to script, then overlay env vars."""
    cfg = dict(DEFAULT_CONFIG)
    cfg_file = script_dir / "okf-lookup.config.json"
    if cfg_file.is_file():
        try:
            raw = json.loads(cfg_file.read_text(encoding="utf-8"))
            for k in DEFAULT_CONFIG:
                if k in raw:
                    cfg[k] = raw[k]
        except (ValueError, OSError) as exc:
            print(f"warning: could not parse {cfg_file}: {exc}", file=sys.stderr)

    # Env-var overlay: OKF_LOOKUP_<KEY_UPPER>=value
    for key in DEFAULT_CONFIG:
        env_key = f"OKF_LOOKUP_{key.upper()}"
        if env_key in os.environ:
            val = os.environ[env_key]
            default = DEFAULT_CONFIG[key]
            if isinstance(default, bool):
                cfg[key] = val.lower() in ("1", "true", "yes", "on")
            elif isinstance(default, int):
                try:
                    cfg[key] = int(val)
                except ValueError:
                    pass
            elif isinstance(default, list):
                cfg[key] = [t.strip() for t in val.split(",") if t.strip()]
            else:
                cfg[key] = val
    return cfg


def tokenize(text: str, cfg: dict) -> set[str]:
    min_len = max(1, int(cfg["min_token_length"]))
    extra = frozenset(cfg.get("extra_stopwords") or [])
    domain = DOMAIN_STOPWORDS if cfg["domain_stopwords_enabled"] else frozenset()
    tokens = TOKEN_RE.findall(text.lower())
    return {
        t for t in tokens
        if len(t) >= min_len
        and t not in STOPWORDS
        and t not in domain
        and t not in extra
    }


def lookup(prompt: str, repo_root: Path, cfg: dict) -> list[tuple[Path, int, str]]:
    index_rel = cfg["index_path"]
    index_path = Path(index_rel) if Path(index_rel).is_absolute() else repo_root / index_rel
    if not index_path.is_file():
        print(f"error: index not found at {index_path}", file=sys.stderr)
        sys.exit(2)

    needle_tokens = tokenize(prompt, cfg)
    if not needle_tokens:
        return []

    index_dir = index_path.parent
    repo_resolved = repo_root.resolve()
    seen: set[Path] = set()
    hits: list[tuple[Path, int, str]] = []
    min_score = int(cfg["min_score"])
    boost_runbooks = bool(cfg["boost_runbooks"])
    include_glossary = bool(cfg["include_glossary"])
    max_preview = int(cfg["max_line_preview"])

    for raw in index_path.read_text(encoding="utf-8", errors="replace").splitlines():
        line_lower = raw.lower()
        score = sum(1 for t in needle_tokens if t in line_lower)
        if score < min_score:
            continue
        if boost_runbooks and "/runbooks/" in line_lower:
            score += 1
        for target in LINK_RE.findall(raw):
            target = target.split("#", 1)[0].strip()
            if not target.endswith(".md"):
                continue
            if target.startswith(("http://", "https://", "mailto:")):
                continue
            if not include_glossary and "/glossary/" in target.lower():
                continue
            resolved = (index_dir / target).resolve()
            try:
                resolved.relative_to(repo_resolved)
            except ValueError:
                continue
            if not resolved.is_file():
                continue
            if resolved in seen:
                continue
            seen.add(resolved)
            trimmed = raw.strip()[:max_preview]
            hits.append((resolved, score, trimmed))

    hits.sort(key=lambda h: -h[1])
    limit = int(cfg["limit"])
    return hits if limit <= 0 else hits[:limit]


def read_prompt(args_prompt: str | None) -> str:
    if args_prompt:
        return args_prompt
    raw = sys.stdin.read()
    stripped = raw.lstrip()
    if stripped.startswith("{"):
        try:
            payload = json.loads(raw)
            return payload.get("prompt") or payload.get("user_prompt") or ""
        except (ValueError, TypeError):
            return raw
    return raw


def main() -> int:
    script_dir = Path(__file__).resolve().parent
    cfg = load_config(script_dir)

    p = argparse.ArgumentParser(
        description="Score knowledge/index.md links by prompt-token overlap.",
        epilog="Config precedence: defaults < scripts/okf-lookup.config.json < OKF_LOOKUP_* env < CLI flags.",
    )
    p.add_argument("prompt", nargs="?", help="Prompt text (else read from stdin).")
    p.add_argument("--root", type=Path, default=None, help="Repo root (default: parent of script dir).")
    p.add_argument("--limit", type=int, default=None, help=f"Top-N (default: {cfg['limit']}, 0=all).")
    p.add_argument("--min-score", type=int, default=None, help=f"Min match score (default: {cfg['min_score']}).")
    p.add_argument("--format", choices=["paths", "reminder"], default=None,
                   help=f"Output format (default: {cfg['format']}).")
    p.add_argument("--no-glossary", action="store_true", help="Exclude knowledge/glossary/ hits.")
    p.add_argument("--boost-runbooks", action="store_true", help="+1 score for runbook hits.")
    p.add_argument("--print-config", action="store_true", help="Print effective config + exit.")
    args = p.parse_args()

    # CLI overrides config
    if args.limit is not None:
        cfg["limit"] = args.limit
    if args.min_score is not None:
        cfg["min_score"] = args.min_score
    if args.format is not None:
        cfg["format"] = args.format
    if args.no_glossary:
        cfg["include_glossary"] = False
    if args.boost_runbooks:
        cfg["boost_runbooks"] = True

    if args.print_config:
        print(json.dumps(cfg, indent=2))
        return 0

    prompt = read_prompt(args.prompt)
    if not prompt.strip():
        return 0  # nothing to surface ? not an error (Claude Code logs exit?0 as hook error)

    root = args.root if args.root else script_dir.parent
    hits = lookup(prompt, root, cfg)

    if not hits:
        return 0  # no matches ? not an error

    repo_resolved = root.resolve()
    if cfg["format"] == "paths":
        for path, _score, _line in hits:
            print(path)
    else:
        print("Relevant knowledge files matched by prompt-token overlap. Read before acting.")
        for path, score, line in hits:
            try:
                rel = path.relative_to(repo_resolved)
                rel_str = str(rel).replace("\\", "/")
            except ValueError:
                rel_str = str(path)
            print(f"- {rel_str} (score={score}) - {line}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/okf-prompt-lookup.ts
#!/usr/bin/env -S npx tsx
// okf-prompt-lookup — score knowledge/index.md lines against a free-form prompt.
// TypeScript port of okf-prompt-lookup.py. Stdlib-only (Node builtins).
// Reads prompt from argv or JSON-stdin (e.g. CC UserPromptSubmit hook).

import * as fs from "node:fs";
import * as path from "node:path";
import * as process from "node:process";
import { fileURLToPath } from "node:url";

const LINK_RE = /\[[^\]]+\]\(([^)]+)\)/g;
const TOKEN_RE = /[a-z0-9][a-z0-9\-_]+/g;

const STOPWORDS = new Set(
  `the a an and or but if then else when while for from to of in on at by with
about into over under across this that these those it its is are was were be
been being have has had do does did will would should could may might must
can shall i me my we us our you your they them their he she his her
please thanks thank yes no ok okay sure can pls plz just want need help
how what why where which who whose make let lets set get got make have take`
    .split(/\s+/)
    .filter(Boolean),
);

const DOMAIN_STOPWORDS = new Set(
  `oriz knowledge agent agents rule rules file files repo repos`
    .split(/\s+/)
    .filter(Boolean),
);

interface Config {
  limit: number;
  min_token_length: number;
  min_score: number;
  max_line_preview: number;
  index_path: string;
  format: "paths" | "reminder";
  extra_stopwords: string[];
  domain_stopwords_enabled: boolean;
  include_glossary: boolean;
  boost_runbooks: boolean;
}

const DEFAULT_CONFIG: Config = {
  limit: 3,
  min_token_length: 3,
  min_score: 1,
  max_line_preview: 120,
  index_path: "knowledge/index.md",
  format: "reminder",
  extra_stopwords: [],
  domain_stopwords_enabled: true,
  include_glossary: true,
  boost_runbooks: false,
};

function loadConfig(scriptDir: string): Config {
  const cfg: Config = { ...DEFAULT_CONFIG };
  const cfgFile = path.join(scriptDir, "okf-lookup.config.json");
  if (fs.existsSync(cfgFile) && fs.statSync(cfgFile).isFile()) {
    try {
      const raw = JSON.parse(fs.readFileSync(cfgFile, "utf-8"));
      for (const k of Object.keys(DEFAULT_CONFIG) as (keyof Config)[]) {
        if (k in raw) {
          (cfg as any)[k] = raw[k];
        }
      }
    } catch (exc) {
      process.stderr.write(`warning: could not parse ${cfgFile}: ${exc}\n`);
    }
  }

  // Env-var overlay: OKF_LOOKUP_<KEY_UPPER>=value
  for (const key of Object.keys(DEFAULT_CONFIG) as (keyof Config)[]) {
    const envKey = `OKF_LOOKUP_${key.toUpperCase()}`;
    if (envKey in process.env) {
      const val = process.env[envKey]!;
      const def = DEFAULT_CONFIG[key];
      if (typeof def === "boolean") {
        (cfg as any)[key] = ["1", "true", "yes", "on"].includes(val.toLowerCase());
      } else if (typeof def === "number") {
        const n = parseInt(val, 10);
        if (!Number.isNaN(n)) (cfg as any)[key] = n;
      } else if (Array.isArray(def)) {
        (cfg as any)[key] = val
          .split(",")
          .map((t) => t.trim())
          .filter(Boolean);
      } else {
        (cfg as any)[key] = val;
      }
    }
  }
  return cfg;
}

function tokenize(text: string, cfg: Config): Set<string> {
  const minLen = Math.max(1, Math.floor(cfg.min_token_length));
  const extra = new Set(cfg.extra_stopwords || []);
  const domain = cfg.domain_stopwords_enabled ? DOMAIN_STOPWORDS : new Set<string>();
  const tokens = text.toLowerCase().match(TOKEN_RE) || [];
  const out = new Set<string>();
  for (const t of tokens) {
    if (
      t.length >= minLen &&
      !STOPWORDS.has(t) &&
      !domain.has(t) &&
      !extra.has(t)
    ) {
      out.add(t);
    }
  }
  return out;
}

interface Hit {
  path: string;
  score: number;
  line: string;
}

function resolvePath(p: string): string {
  return path.resolve(p);
}

function isSubPath(child: string, parent: string): boolean {
  const rel = path.relative(parent, child);
  return !rel.startsWith("..") && !path.isAbsolute(rel);
}

function lookup(prompt: string, repoRoot: string, cfg: Config): Hit[] {
  const indexRel = cfg.index_path;
  const indexPath = path.isAbsolute(indexRel)
    ? indexRel
    : path.join(repoRoot, indexRel);
  if (!fs.existsSync(indexPath) || !fs.statSync(indexPath).isFile()) {
    process.stderr.write(`error: index not found at ${indexPath}\n`);
    process.exit(2);
  }

  const needleTokens = tokenize(prompt, cfg);
  if (needleTokens.size === 0) return [];

  const indexDir = path.dirname(indexPath);
  const repoResolved = resolvePath(repoRoot);
  const seen = new Set<string>();
  const hits: Hit[] = [];
  const minScore = Math.floor(cfg.min_score);
  const boostRunbooks = Boolean(cfg.boost_runbooks);
  const includeGlossary = Boolean(cfg.include_glossary);
  const maxPreview = Math.floor(cfg.max_line_preview);

  const content = fs.readFileSync(indexPath, "utf-8");
  const lines = content.split(/\r?\n/);

  for (const raw of lines) {
    const lineLower = raw.toLowerCase();
    let score = 0;
    for (const t of needleTokens) {
      if (lineLower.includes(t)) score += 1;
    }
    if (score < minScore) continue;
    if (boostRunbooks && lineLower.includes("/runbooks/")) score += 1;

    // Reset regex state before each line
    LINK_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = LINK_RE.exec(raw)) !== null) {
      let target = m[1].split("#", 1)[0].trim();
      if (!target.endsWith(".md")) continue;
      if (
        target.startsWith("http://") ||
        target.startsWith("https://") ||
        target.startsWith("mailto:")
      )
        continue;
      if (!includeGlossary && target.toLowerCase().includes("/glossary/")) continue;

      const resolved = resolvePath(path.join(indexDir, target));
      if (!isSubPath(resolved, repoResolved) && resolved !== repoResolved) continue;
      if (!fs.existsSync(resolved) || !fs.statSync(resolved).isFile()) continue;
      if (seen.has(resolved)) continue;
      seen.add(resolved);
      const trimmed = raw.trim().slice(0, maxPreview);
      hits.push({ path: resolved, score, line: trimmed });
    }
  }

  hits.sort((a, b) => b.score - a.score);
  const limit = Math.floor(cfg.limit);
  return limit <= 0 ? hits : hits.slice(0, limit);
}

function readStdin(): string {
  try {
    return fs.readFileSync(0, "utf-8");
  } catch {
    return "";
  }
}

function readPrompt(argsPrompt: string | undefined): string {
  if (argsPrompt) return argsPrompt;
  const raw = readStdin();
  const stripped = raw.replace(/^\s+/, "");
  if (stripped.startsWith("{")) {
    try {
      const payload = JSON.parse(raw);
      return payload.prompt || payload.user_prompt || "";
    } catch {
      return raw;
    }
  }
  return raw;
}

interface Args {
  prompt?: string;
  root?: string;
  limit?: number;
  min_score?: number;
  format?: "paths" | "reminder";
  no_glossary: boolean;
  boost_runbooks: boolean;
  print_config: boolean;
}

function parseArgs(argv: string[]): Args {
  const args: Args = {
    no_glossary: false,
    boost_runbooks: false,
    print_config: false,
  };
  const positional: string[] = [];
  let i = 0;
  while (i < argv.length) {
    const a = argv[i];
    if (a === "--root") {
      args.root = argv[++i];
    } else if (a === "--limit") {
      args.limit = parseInt(argv[++i], 10);
    } else if (a === "--min-score") {
      args.min_score = parseInt(argv[++i], 10);
    } else if (a === "--format") {
      const v = argv[++i];
      if (v !== "paths" && v !== "reminder") {
        process.stderr.write(`error: --format must be 'paths' or 'reminder'\n`);
        process.exit(2);
      }
      args.format = v;
    } else if (a === "--no-glossary") {
      args.no_glossary = true;
    } else if (a === "--boost-runbooks") {
      args.boost_runbooks = true;
    } else if (a === "--print-config") {
      args.print_config = true;
    } else if (a === "-h" || a === "--help") {
      process.stdout.write(
        "Usage: okf-prompt-lookup [prompt] [--root DIR] [--limit N] [--min-score N] [--format paths|reminder] [--no-glossary] [--boost-runbooks] [--print-config]\n",
      );
      process.exit(0);
    } else if (a.startsWith("--")) {
      process.stderr.write(`error: unknown flag ${a}\n`);
      process.exit(2);
    } else {
      positional.push(a);
    }
    i++;
  }
  if (positional.length > 0) args.prompt = positional.join(" ");
  return args;
}

function main(): number {
  const scriptFile = fileURLToPath(import.meta.url);
  const scriptDir = path.dirname(scriptFile);
  const cfg = loadConfig(scriptDir);

  const args = parseArgs(process.argv.slice(2));

  // CLI overrides config
  if (args.limit !== undefined) cfg.limit = args.limit;
  if (args.min_score !== undefined) cfg.min_score = args.min_score;
  if (args.format !== undefined) cfg.format = args.format;
  if (args.no_glossary) cfg.include_glossary = false;
  if (args.boost_runbooks) cfg.boost_runbooks = true;

  if (args.print_config) {
    process.stdout.write(JSON.stringify(cfg, null, 2) + "\n");
    return 0;
  }

  const prompt = readPrompt(args.prompt);
  if (!prompt.trim()) return 0; // nothing to surface — not an error

  const root = args.root ? path.resolve(args.root) : path.dirname(scriptDir);
  const hits = lookup(prompt, root, cfg);

  if (hits.length === 0) return 0; // no matches — not an error

  const repoResolved = resolvePath(root);
  if (cfg.format === "paths") {
    for (const h of hits) process.stdout.write(h.path + "\n");
  } else {
    process.stdout.write(
      "Relevant knowledge files matched by prompt-token overlap. Read before acting.\n",
    );
    for (const h of hits) {
      let relStr: string;
      if (isSubPath(h.path, repoResolved) || h.path === repoResolved) {
        relStr = path.relative(repoResolved, h.path).replace(/\\/g, "/");
      } else {
        relStr = h.path;
      }
      process.stdout.write(`- ${relStr} (score=${h.score}) - ${h.line}\n`);
    }
  }
  return 0;
}

process.exit(main());
